x/staking/types/validator.go

```diff
@@ -421,7 +422,18 @@ func (v Validator) GetMoniker() string          { return v.Description.Moniker }
 func (v Validator) GetStatus() sdk.BondStatus   { return v.Status }
 func (v Validator) GetOperator() sdk.ValAddress { return v.OperatorAddress }
 func (v Validator) GetConsPubKey() crypto.PubKey {
-	return sdk.MustGetPubKeyFromBech32(sdk.Bech32PubKeyTypeConsPub, v.ConsensusPubkey)
+	// The way things are refactored now, v.ConsensusPubkey is sometimes a TM
+	// ed25519 pubkey, sometimes our own ed25519 pubkey. This is very ugly and
+	// inconsistent.
+	// Luckily, here we coerce it into a TM ed25519 pubkey always, as this
+	// pubkey will be passed into TM.
+	pk := sdk.MustGetPubKeyFromBech32(sdk.Bech32PubKeyTypeConsPub, v.ConsensusPubkey)
+
+	if intoTmPk, ok := pk.(cryptotypes.IntoTmPubKey); ok {
+		return intoTmPk.AsTmPubKey()
+	}
```

Comment:
yeah, it's ugly. A follow-up (big) refactor is to remove all `crypto.PubKey` (TM's PubKey interface), and use our own `cryptotypes.PubKey`.

Reply:
Why do we need to do a conversion? Our keys support `tendermint PubKey` interface.

Reply:
As described in the comment, this is because returned value from `GetConsPubKey()` can then be passed to function from tendermint (see `encoding.PubKeyToProto` in this same file) which expects a tendermint pubkey https://github.com/tendermint/tendermint/blob/e8eef856196697c803ac54fa625d83fb276d5977/crypto/encoding/codec.go#L12

Reply:
in the link you provided, it's an interface. And our implementation works in the same way. The idea of using interfaces is to avoid this conversions if the functionality is the same.

My point is to know if we need of if we can use our interface implementation whenever possible.

Reply:
yeah but if you look at the implementation, then it's expecting tm ed25519 https://github.com/tendermint/tendermint/blob/e8eef856196697c803ac54fa625d83fb276d5977/crypto/encoding/codec.go#L15 else returns an error that the type is not supported.

Reply:
Ah, this is bad I  think. OK. Maybe we should sort out this dependencies later on.
@marbar3778  -- what do you think? Why tendermint is doing this hard assumption for type, rather than using interfaces ? We should probably have a method and a separate interface for that, eg `ProtoPubKeyExporter`

Reply:
Copying Marko comment here:
> In Tendermint we are currently only support ed25519 for this release. This was done in order to reduce the scope of work because we thought the release would be a few months ago. We will rework this part of Tendermint in the future to not have the type hardcoded.

> As for the going to Tendermint key and sdk key, I am a bit scared of this design but it works for now. The dependency of Tendermint will change quite a bit in the coming releases.